src/Simplex/Messaging/Agent.hs

```diff
@@ -1085,13 +1122,39 @@ enqueueMessage c cData@ConnData {connId} sq msgFlags aMessage = do
       liftIO $ createSndMsg db connId msgData
       liftIO $ createSndMsgDelivery db connId sq internalId
       pure internalId
+    processResults :: ((EIORef AgentMsgId, (ConnData, NonEmpty SndQueue, MsgFlags, AMessage)), Either AgentErrorType InternalId) -> m (Maybe (ConnData, [SndQueue], AgentMsgId))
+    processResults ((r, (cData, sq :| sqs, _, _)), mId_) = case mId_ of
+      Left e -> Nothing <$ writeIORef r (Left e)
+      Right mId -> do
+        let InternalId msgId = mId
+        writeIORef r $ Right msgId
+        queuePendingMsgs c sq [mId]
+        let sqs' = filter isActiveSndQ sqs
+        pure $ if null sqs' then Nothing else Just (cData, sqs', msgId)
+
+enqueueSavedMessage :: AgentMonad' m => AgentClient -> ConnData -> AgentMsgId -> SndQueue -> m ()
+enqueueSavedMessage c cData msgId sq = enqueueSavedMessageB c [(cData, [sq], msgId)]
+
+enqueueSavedMessageB :: AgentMonad' m => AgentClient -> [(ConnData, [SndQueue], AgentMsgId)] -> m ()
+enqueueSavedMessageB c reqs = do
+  -- saving to the database moved to the start to avoid race conditions when delivery is read from queue before it is saved
```

Comment:
strange to have this comment in code - "moved to the start" with comparison to what? It's clear in PR, but confusing in code. Better to delete it

Reply:
"store batch must be completed before starting delivery due to read-after-write hazard in <name-where-the-read-happens>" ?